// wombo-dream-api/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Types = exports.buildDefaultInstance = exports.GoogleAuthentifier = exports.WomboDream = exports.DEFAULT_AUTHENTIFICATION_KEY = exports.DEFAULT_UPLOAD_URL = exports.DEFAULT_GALLERY_API_SUFFIX = exports.DEFAULT_STYLE_API_SUFFIX = exports.DEFAULT_SHOP_API_SUFFIX = exports.DEFAULT_TASK_API_SUFFIX = exports.DEFAULT_TASK_API_URL = exports.DEFAULT_ORIGIN_URL = void 0;
var WomboDream_1 = __importDefault(require("./WomboDream"));
var GoogleAuthentifier_1 = __importDefault(require("./GoogleAuthentifier"));
exports.DEFAULT_ORIGIN_URL = 'https://app.wombo.art';
exports.DEFAULT_TASK_API_URL = 'https://paint.api.wombo.ai/api/%(suffix)s';
exports.DEFAULT_TASK_API_SUFFIX = 'tasks/%(taskId)s';
exports.DEFAULT_SHOP_API_SUFFIX = 'shop/%(taskId)s';
exports.DEFAULT_STYLE_API_SUFFIX = 'styles/';
exports.DEFAULT_GALLERY_API_SUFFIX = 'gallery/%(taskId)s';
exports.DEFAULT_UPLOAD_URL = 'https://mediastore.api.wombo.ai/io/';
exports.DEFAULT_AUTHENTIFICATION_KEY = 'AIzaSyDCvp5MTJLUdtBYEKYWXJrlLzu1zuKM6Xw';
exports.WomboDream = __importStar(require("./WomboDream"));
exports.GoogleAuthentifier = __importStar(require("./GoogleAuthentifier"));
var buildDefaultInstance = function (email, password) {
    var wombo = new WomboDream_1.default(new GoogleAuthentifier_1.default(exports.DEFAULT_AUTHENTIFICATION_KEY, email, password), exports.DEFAULT_TASK_API_URL, exports.DEFAULT_TASK_API_SUFFIX, exports.DEFAULT_SHOP_API_SUFFIX, exports.DEFAULT_STYLE_API_SUFFIX, exports.DEFAULT_GALLERY_API_SUFFIX, exports.DEFAULT_ORIGIN_URL, exports.DEFAULT_UPLOAD_URL);
    return wombo;
};
exports.buildDefaultInstance = buildDefaultInstance;
exports.Types = __importStar(require("./types"));
